File: src/body.go
package src

// Physical Body

import (
	"math"

	raylib "github.com/gen2brain/raylib-go/raylib"
)

type Body struct {
	Mass          float64
	Width         float64
	Height        float64
	Location      Vector
	Vel           Vector
	Acc           Vector
	RotationAngle float64
}

func NewBody(x, y, mass float64) *Body {
	return &Body{Mass: mass, Width: mass * 4, Height: mass * 4, Location: Vector{X: x, Y: y}}
}

// Drawing Function
func (body *Body) Draw() {
	DrawRotationalRect(&body.Location, body.Width, body.Height, body.RotationAngle)
}

// Updating Body State
func (body *Body) Update() {
	body.HandleForceStacking()
	body.Keys()
}

// Handle Multiple Forces on Body
func (body *Body) HandleForceStacking() {
	body.Vel.Limit(10)
	body.Location.Add(body.Vel)
	body.Vel.Add(body.Acc)
	body.Acc.Multi(0)
}

// Apply Force To The Body
func (body *Body) ApplyForce(force Vector) {
	body.Acc.Add(*Vector_Div(force, body.Mass))
}

// Attract other Body
func (body *Body) Attract(other *Body) {
	dir := SubVectors(body.Location, other.Location)
	distSQ := math.Max(math.Min(dir.Mag()*dir.Mag(), 25), 2500)
	strength := body.Mass * other.Mass * 0.3 / distSQ
	dir.Multi(strength)
	other.ApplyForce(*dir)
}

// Display Velocity Direction
func (body *Body) VelDirection() {
	vel_dir := AddVectors(body.Location, *Vector_Multi(body.Vel, 12))
	raylib.DrawLine(
		int32(
			body.Location.X,
		),
		int32(body.Location.Y),
		int32(vel_dir.X),
		int32(vel_dir.Y),
		raylib.Black,
	)
}

// Handling Keyboard Keyboard Input
func (body *Body) Keys() {
	if raylib.IsKeyDown(raylib.KeySpace) {
		body.RotationAngle += (math.Pi / 180) * 2
	}
	if raylib.IsKeyDown(raylib.KeyBackspace) {
		body.RotationAngle -= (math.Pi / 180) * 2
	}
	if raylib.IsKeyDown(raylib.KeyD) {
		body.ApplyForce(*NewVector(5, 0))
	} else if raylib.IsKeyDown(raylib.KeyA) {
		body.ApplyForce(*NewVector(-5, 0))
	}
	if raylib.IsKeyDown(raylib.KeyS) {
		body.ApplyForce(*NewVector(0, 5))
	} else if raylib.IsKeyDown(raylib.KeyW) {
		body.ApplyForce(*NewVector(0, -5))
	}
}

File: src/body.cpp
#include "../lib/raylib/raylib.h"
#include "src.h"
#include <cmath>

// Body Constructor
Body::Body(double _x, double _y, double _mass, double _angle) {
  this->mass = _mass;
  this->location.x = _x;
  this->location.y = _y;
  this->angle = _angle;
  this->size.x = this->mass * 2;
  this->size.y = this->mass;
}

// Updating body state
void Body::update() {
  this->handle_force_stacking();
  this->handle_input();
  // this->angle = std::atan2(this->vel.y, this->vel.x);
}

// Halding multiple forces on body
void Body::handle_force_stacking() {
  this->vel.limit(10);
  this->location.add(this->vel);
  this->vel.add(this->acc);
  this->acc.multi(0);
}

// Apply force to the body
void Body::apply_force(Vector &force) {
  Vector force_acc = Vector::Vector_div(force, this->mass);
  this->acc.add(force_acc);
}

// Show velocity directions
void Body::vel_direction() {
  Vector vel_vec = Vector::Vector_multi(this->vel, 8);
  Vector vel_dir = Vector::AddVectors(this->location, vel_vec);
  DrawLine(this->location.x, this->location.y, vel_dir.x, vel_dir.y, RED);
}

// Handle keyboard input
void Body::handle_input() {
  if (IsKeyDown(KEY_H)) {
    this->angle += (M_PI / 180) * 2;
  } else if (IsKeyDown(KEY_L)) {
    this->angle -= (M_PI / 180) * 2;
  }

  if (IsKeyDown(KEY_SPACE)) {
    Vector force =
        Vector(10 * std::cos(this->angle), 10 * std::sin(this->angle));
    this->apply_force(force);
  }

  if (IsKeyDown(KEY_D)) {
    Vector force = Vector(10, 0);
    this->apply_force(force);
  } else if (IsKeyDown(KEY_A)) {
    Vector force = Vector(-10, 0);
    this->apply_force(force);
  }
  if (IsKeyDown(KEY_S)) {
    Vector force = Vector(0, 10);
    this->apply_force(force);
  } else if (IsKeyDown(KEY_W)) {
    Vector force = Vector(0, -10);
    this->apply_force(force);
  }
}
